ambry-router/src/main/java/com.github.ambry.router/GetBlobOperation.java

```
@@ -0,0 +1,777 @@
+      FutureResult<ReadableStreamChannel> futureResult, Callback<ReadableStreamChannel> callback,
+      OperationCompleteCallback operationCompleteCallback, Time time)
+      throws RouterException {
+    super(routerConfig, clusterMap, responseHandler, blobIdStr, futureResult, callback, time);
+    this.operationCompleteCallback = operationCompleteCallback;
+    this.blobIdFactory = new BlobIdFactory(clusterMap);
+    firstChunk = new FirstGetChunk(blobId);
+  }
+
+  /**
+   * {@inheritDoc}
+   * <br>
+   * If the callback for the operation was already called, then abort the reads from the channel and the caller
+   * will be notified as part of the read callback eventually. If not, the notification will happen as part of the
+   * operation completion callback.
+   */
+  @Override
+  void abort(Exception abortCause) {
+    if (operationCallbackInvoked.compareAndSet(false, true)) {
+      operationCompleteCallback.completeOperation(operationFuture, operationCallback, null, abortCause);
+    } else {
+      operationException.set(abortCause);
+      if (readCalled) {
+        completeRead();
+      }
+    }
+    operationCompleted = true;
+  }
+
+  /**
+   * Return the {@link MessageFormatFlags} to associate with a getBlob operation.
+   * @return {@link MessageFormatFlags#Blob}
+   */
+  @Override
+  MessageFormatFlags getOperationFlag() {
+    return MessageFormatFlags.Blob;
+  }
+
+  /**
+   * Do all that needs to be done (cleanup, notification, etc.) on chunk completion and mark the state of the chunk
+   * appropriately.
+   * @param chunk the chunk that has completed.
+   */
+  private void onChunkOperationComplete(GetChunk chunk) {
+    if (chunk.getChunkException() != null) {
+      // if operation callback was already called, then this exception will have to be notified as part of the
+      // read callback.
+      setOperationException(chunk.getChunkException());
+    }
+    if (chunk == firstChunk) {
```

I don't think there is a correctness issue here, but I think we should fail faster like you suggest. Will make the change.